src/security/ServerOptions.cc

```diff
@@ -390,12 +396,12 @@ Security::ServerOptions::loadDhParams()
     parsedDhParams.resetWithoutLocking(dhp);
```

Comment:
> > PR: Squid https_port ... tls-dh=curve:dhparams.pem configuration is
> > supposed to support both ECDHE and FFDHE key exchange mechanisms
> 
> [Amos](https://github.com/squid-cache/squid/pull/1949#pullrequestreview-2446687296): This is false. Squid has never supported FFDHE (RFC 7919). Squid implements RFC 4492. That only supports classical DH cipher, or ECDH cipher(s).

I assume this is a disagreement on _terminology_ (i.e. "FFDHE" vs "classical DH") rather than on the primary point of that statement (i.e. "supposed to support _both_") that this PR is based on. @yadij, please correct me if this assumption is wrong.

FWIW, I am OK with any correct terminology, but I like FFDHE term better because, unlike "classical DH", FFDHE is pretty well defined ([RFC 7919 Section 1.2](https://www.rfc-editor.org/rfc/rfc7919#section-1.2)) and because @sp-andwei [tests](https://github.com/squid-cache/squid/pull/1949#issue-2672465005) with ffdhe4096 named group appear to demonstrate that at least one FFDHE mechanism _is_ supported by Squid. @yadij, I assume that you are not disputing the correctness of those tests, but please correct me if my assumption is wrong.

Reply:
As far as I understand, FFDHE and DHE are used interchangably [(RFC 7919):](https://datatracker.ietf.org/doc/html/rfc7919#section-1.2). 

>    The terms "DHE" or "FFDHE" are used in this document to refer to the
>    finite-field-based Diffie-Hellman ephemeral key exchange mechanism in
>    TLS. 

Squid definitely supported DHE in the past via the `tls-dh` parameter.

Reply:
As additional comment to your inital statement:


- `nil:PC`
  - That case never worked and does not currently work in `v6`/`master`:


- `CX:PCX` or `CX:PCY`
  - That case have never worked in a sense that any additional parameters have been loaded and used and does not currently work in that sense in the `v6` or `master` branches. They *do* work if a valid name for the curve is specified, because the curve-to-be-used is always set by name, see my long post above:

- `nil:PD` and (missing above) `D:PD`
  - (addressed by PR) both broken in `v6`/`master`, used to work in (`v4`/`v5`); loading of PD successful, but loading it into `SSL_CTX` ist not

- `C:PD`
  - addressed by PR, used used to work in `v4`/`v5`, with `v6`/`master` throws errors and only configures ECDHE successfully
 
- `C:nil` and `nil:nil` 
  - those two should work like expected

Reply:
FYI: Tested my patch with openssl 3.0, squid v6 + patch 
- `tls-dh=<path>/dhparams4096.pem`: connecting with `openssl s_client` defaults: TLS-1.3 with `Server Temp Key: X25519, 253 bits`
  - as in this case `SSL_CTX_set1_groups` is not called and I assume openssl just falls back to some default behavior (using curve `X25519` in this case)
  - DHE works, too, when forcing it using TLS-1.2 and an explicit DHE cihper: `openssl s_client -tls1_2 -cipher DHE-RSA-AES256-SHA256 -connect rproxytest.intern01.local:8443` --> `Server Temp Key: DH, 4096 bits`

- `tls-dh=ffdhe4096:<path>/dhparams4096.pem`
  - with `openssl s_client` defaults TLS-1.3 with `Server Temp Key: DH, 4096 bits`
  - with `openssl s_client -tls1_2 -cipher DHE-RSA-AES256-SHA256 -connect rproxytest.intern01.local:8443` --> `Server Temp Key: DH, 4096 bits`
  - forcing ECDHE: `openssl s_client -tls1_2 -cipher ECDHE-RSA-AES256-SHA384 -connect rproxytest.intern01.local:8443` --> handshake failure (as would be expected)

Reply:
1. I added some confusion by bringing squid v5 into that. I got the commit that added the openssl < 3 `#if` wrong, squid v5 uses basically the same code as squid v6. Sorry about that.

Some more verifying:
```sh
openssl ecparam -name brainpoolP256r1 -out secp384r1.pem
openssl ecparam -name prime256v1 -out prime256v1.pem

openssl ecparam -in secp384r1.pem -text -noout
ASN1 OID: secp384r1
NIST CURVE: P-384

openssl ecparam -in prime256v1.pem -text -noout
ASN1 OID: prime256v1
NIST CURVE: P-256

```

2. `CX:PCY`, loading of `PCY` does *not* work with the PR. Neither did it work with squid v4, v5, v6, master. `CX` is used and that's it.
  - `tls-dh=prime256v1:<path>/secp384r1.pem`
    - v4: `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: ECDH, prime256v1, 256 bits`
    - v6: `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: ECDH, prime256v1, 256 bits`
  - `tls-dh=secp384r1:<path>/prime256v1.pem`
    -  v4: `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: ECDH, secp384r1, 384 bits`
    -  v6: `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: ECDH, secp384r1, 384 bits`
3. `nil:PCY`, that case does *not* work with the PR. Neither did it work with squid v4, v5, v6, master. openssl defaults are used (currently, `X25519` curve for ECDHE ist installed)
  - `tls-dh=`<path>/prime256v1.pem`
    - v4: `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: X25519, 253 bits`
    - v6: `WARNING: Failed to decode DH parameters '/rootdisk/config/prime256v1.pem'; `openssl s_client -connect rproxytest.intern01.local:8443` --> `Server Temp Key: X25519, 253 bits`

The time it took to verify those combinations would have probably been enough to implement some logic to allow specifying a list of groups... 

I just wanted to fix that *one case* relevant for us that has been broken going from v4 to v6. I think I have demonstrated that my quite minimalistic PR does exactly that and does not alter or worsen the behavior in any other way.

Reply:
> 1. I added some confusion by bringing squid v5 into that. I got the commit that added the openssl < 3 `#if` wrong, squid v5 uses basically the same code as squid v6. Sorry about that.

Thank you for checking _and_ keeping PR description up to date!
 
> Some more verifying: ...
> 2. `CX:PCY`, loading of PCY does not work with PR, ... v4, v5, v6, master. CX is used and that's it.
> 3. `nil:PCY`, does not work with PR, ... v4, v5, v6, master. openssl defaults are used

Thank you for testing these. Both outcomes appear to support PR's assertion that parameters file is unusable for configuring EC curves (before and after PR changes).

----

The speculations below record my understanding of _why_ the above use cases are not (and have not been) supported. You do _not_ have to validate these speculations, but I would appreciate if you can do that (without heroic efforts) for my own edification.

Also, if speculations below are correct, then I wonder whether this PR can avoid specifying parameters key type (i.e. "DH") and let OpenSSL decide what that parameter file contains. Could something like that work and add support for `CX:PCY` and `nil:PCY` cases while removing one more hard-coded string/decision from Squid code?? OSSL_DECODER_CTX_new_for_pkey documentation looks encouraging, but there may be some post-call caveats I do not know about:

```
The search of decoder implementations can also be limited with keytype and selection, which specifies the
expected resulting keytype and contents.  NULL and zero are valid and signify that the decoder
implementations will find out the keytype and key contents on their own from the input they get.
```


I speculate that official code does not support `CX:PCY` case because, as already explained in this PR description, official code calls SSL_CTX_set_tmp_dh() with the wrong parameter type (EVP_PKEY instead of DH). PR code does not support this use case because it sets parameters file type to "DH" instead of "EC".

I suspect `nil:PCY` case does not work because all code versions set parameters file type to to "DH" (instead of "EC") in this case.

Reply:
> (without heroic efforts)

touché ;)

> official code calls SSL_CTX_set_tmp_dh() 

That is the main reason why this never has worked and also does not work in `v5`/`v6` to load any additional ECDHE parameters (which I have yet to actually encounter in the wild). The ECDHE-groups to use were always and are now successfully set by `SSL_CTX_set_tmp_ecdh` or `SSL_CTX_set1_groups` earlier in the code. But the parameters for those are solely derived from the name given before the colon, i.e., `CX`.

TBH, this is a probably a rather minor thing on the whole, but we did use the option to specify a curve *and* some locally generated dh-groups (using the DHE parameter loading). So that configuration option is "out there" and it would be nice if it worked (especially since loading DH parameters does no longer work at all currently, due to the `DH*` vs `EVP_PKEY` issue).

Reply:
> I guess it could work. 

Great, it sounds like we are more-or-less on the same page.

> But then again, it would probably make more sense to be able to specify a list of group/param pairs

Yes, perhaps something like `tls-dh-groups=<group1>,<group2>,...` where each entry uses either `ec-curve:<name>` or `file:paramters.pem` format. I was just hoping that we could allow folks that want to use files for all groups to use parameters.pem entries to specify EC curve names (instead of using a special syntax for them).

Any such enhancement is outside _this_ PR scope, of course!

> I still fail to see how you would go about providing any custom EC-params

I assume they have no practical value in Squid deployments. If the overall configuration approach is flexible enough, then we can add code to support them if/when there is a real use case (important enough to warrant that extra work). Until then, naming EC curve using parameters file could be helpful in some situations (as mentioned above), but even that is not essential.